### PR TITLE
extensionrepo: handle ref changes correctly

### PR DESCRIPTION
**Pull request by nicks**
Hello @milas, @landism,

Please review the following commits I made in branch nicks/ref2:

1d09bba3318f59fabe6a4f3e56a4edd373608299 (2022-02-18 13:49:08 -0500)
extensionrepo: handle ref changes correctly
Fixes https://github.com/tilt-dev/tilt/issues/5508

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Review by landism — COMMENTED** [review]
This seems a little confusing at the moment. IIUC:
1. If you `tilt up` with a repo pointing at one ref, and then change the ref, we [delete the repo and re-fetch it](https://github.com/tilt-dev/tilt/blob/449fe2dbcc77c7bfacea82d1b0e14f93c4b0194a/internal/controllers/core/extensionrepo/reconciler.go#L121-L128)
2. This PR only takes effect if you `tilt up` with a repo pointing at one ref, stop tilt, then `tilt up` with the same repo pointing at a different ref?

for more context on my confusion: I tried testing the existing behavior when changing a ref in a Tiltfile, since [this](https://github.com/tilt-dev/tilt/blob/449fe2dbcc77c7bfacea82d1b0e14f93c4b0194a/internal/controllers/core/extensionrepo/reconciler.go#L216-L221) looked like it'd be a bug, except we already handle that case [upstream](https://github.com/tilt-dev/tilt/blob/449fe2dbcc77c7bfacea82d1b0e14f93c4b0194a/internal/controllers/core/extensionrepo/reconciler.go#L122-L125)

1. Would the problem solved by this PR be more simply (though less efficiently) solved by changing [this `&&`](https://github.com/tilt-dev/tilt/blob/449fe2dbcc77c7bfacea82d1b0e14f93c4b0194a/internal/controllers/core/extensionrepo/reconciler.go#L123) to a `||`, so that we also re-fetch on first `tilt up`?
2. Alternatively, would it make sense to get rid of that if statement and instead have the new code handle the spec check?

**Comment by nicks** [comment]
nope! you're misunderstanding this PR. it's for handling a variety of cases where the on-disk repo can't pull updates, so we delete the repo to re-pull. for details, probably better to read:

1) the attached issue, or
2) the changes to the test fixture

File: internal/controllers/core/extensionrepo/reconciler.go [hunk 1]
```diff
@@ -229,26 +229,47 @@ func (r *Reconciler) reconcileDownloaderRepo(ctx context.Context, state *repoSta
...
+	if exists && state.spec.Ref != "" {
+		// If an explicit ref is specified, there's no reason to pull a new version.
+		err := r.dlr.RefSync(importPath, state.spec.Ref)
+		if err == nil {
+			needsDownload = false
+		} else {
+			// TODO(nick): The more efficient thing to do here would be to
+			// checkout the main branch and pull. But I don't think this case will
+			// happen very often, so it's safer to delete and re-download.
+			_ = os.RemoveAll(destPath)
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/controllers/core/extensionrepo/reconciler.go]: done

File: internal/controllers/core/extensionrepo/reconciler.go [hunk 2]
```diff
@@ -229,26 +229,47 @@ func (r *Reconciler) reconcileDownloaderRepo(ctx context.Context, state *repoSta
...
-	if err != nil {
-		state.status = v1alpha1.ExtensionRepoStatus{Error: fmt.Sprintf("verifying download destination: %v", err)}
-		return ctrl.Result{}
-	}
+	if needsDownload {
+		_, err = r.dlr.Download(importPath)
+		if err != nil {
+			if exists {
+				// Delete any partial state.
+				_ = os.RemoveAll(destPath)
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/controllers/core/extensionrepo/reconciler.go]: actually better to fall into the other error

**Review by landism — APPROVED** [review]

**Pull request merged**